### ISSUE TITLE
Crash in LuaJIT parser when build with ASan

### ISSUE THREAD

**Issue by Buristan** [tarantool/tarantool#5880]
If one builds Tarantool with the following command:
```bash
$ CC=clang CXX=clang++ cmake . -DCMAKE_BUILD_TYPE=Debug -DENABLE_WERROR=ON -DENABLE_ASAN=ON -DENABLE_UB_SANITIZER=ON -DENABLE_FUZZER=ON  && make -j
```
and runs the following Lua chunk, it is lead to crash:
```bash
$ ASAN=ON LSAN_OPTIONS=suppressions=asan/lsan.supp ASAN_OPTIONS=heap_profile=0:unmap_shadow_on_exit=1:detect_invalid_pointer_pairs=1:symbolize=1:detect_leaks=1:dump_instruction_bytes=1:print_suppressions=0 src/tarantool -e 'local s = "local a; a="..string.rep("{", 400) loadstring(s) os.exit()'
Segmentation fault
```

I've run this command inside gdb to get backtrace and it is the following:
```
#0  0x000000000192bb75 in lex_scan (ls=0x7fffe99fefa0, tv=0x7fffe99fefb0) at lj_lex.c:291
#1  0x000000000192b7af in lj_lex_next (ls=0x7fffe99fefa0) at lj_lex.c:454
#2  0x0000000001943716 in lex_check (ls=0x7fffe99fefa0, tok=123) at lj_parse.c:1000
#3  0x0000000001947c20 in expr_table (ls=0x7fffe99fefa0, e=0x7fffe9983780) at lj_parse.c:1714
#4  0x000000000194652d in expr_simple (ls=0x7fffe99fefa0, v=0x7fffe9983780) at lj_parse.c:2004
#5  0x0000000001944360 in expr_unop (ls=0x7fffe99fefa0, v=0x7fffe9983780) at lj_parse.c:2077
#6  0x0000000001943e1f in expr_binop (ls=0x7fffe99fefa0, v=0x7fffe9983780, limit=0) at lj_parse.c:2090
#7  0x0000000001943787 in expr (ls=0x7fffe99fefa0, v=0x7fffe9983780) at lj_parse.c:2109
#8  0x0000000001948385 in expr_table (ls=0x7fffe99fefa0, e=0x7fffe9984520) at lj_parse.c:1734
#9  0x000000000194652d in expr_simple (ls=0x7fffe99fefa0, v=0x7fffe9984520) at lj_parse.c:2004
#10 0x0000000001944360 in expr_unop (ls=0x7fffe99fefa0, v=0x7fffe9984520) at lj_parse.c:2077
...
#740 0x0000000001a34d5b in lj_BC_FUNCC () at /home/s.kaplun/builds/tarantool/asant/third_party/luajit/src/lj_vm.S:811
#741 0x0000000001860ea1 in lua_pcall (L=0x40000378, nargs=0, nresults=0, errfunc=0) at lj_api.c:1158
#742 0x000000000158b45f in luaT_call (L=0x40000378, nargs=0, nreturns=0) at lua/utils.c:1022
#743 0x0000000001557022 in run_script_f (ap=0x7fffe9800448) at lua/init.c:613
#744 0x0000000000565a03 in fiber_cxx_invoke(int (*)(__va_list_tag*), __va_list_tag*) (f=0x15558c0 <run_script_f>, ap=0x7fffe9800448) at lib/core/fiber.h:882
#745 0x000000000165e158 in fiber_loop (data=0x0) at fiber.c:879
#746 0x0000000002499a87 in coro_init () at third_party/coro/coro.c:110
#747 0x0000000000000000 in ?? ()

```

Looks like there is no enough stack for ASan to do it work:
```
(gdb) x /i $rip
=> 0x192bb75 <lex_scan+549>:    call   0x546240 <__sanitizer_cov_trace_const_cmp4>
(gdb) f 0
#0  0x000000000192bb75 in lex_scan (ls=0x7fffe99fefa0, tv=0x7fffe99fefb0) at lj_lex.c:291
291         if (lj_char_isident(ls->c)) {
(gdb) p $rbp
$8 = (void *) 0x7fffe9982870
(gdb) f 746
#746 0x0000000002499a87 in coro_init () at third_party/coro/coro.c:110
110       func ((void *)arg);
(gdb) p $rbp
$5 = (void *) 0x7fffe99ffff0
(gdb) p (0x7fffe99ffff0 - 0x7fffe9982870)/1024
$7 = 501
```

C stack for the first Tarantool fiber is approximately limited by 500 KB.

The crash happens at the 146-th syntax level:
```
(gdb) p ls->level
$9 = 146
```
The limit for the LuaJIT is 200 (`LJ_MAX_XLEVEL`), see <lj_def.h>.

**Comment by igormunkin** [comment]
I tried your recipe on `3.0.0-alpha1-23-ge78afb3bf` (e78afb3bf) without any changes. Segmentation fault still occurs.

```
$ CC=clang-11 CXX=clang++-11 cmake . -DCMAKE_BUILD_TYPE=Debug -DENABLE_WERROR=ON -DENABLE_ASAN=ON -DENABLE_UB_SANITIZER=ON -DENABLE_FUZZER=ON
<...>
$ cmake --build . --parallel 8
<...>
$ ASAN=ON LSAN_OPTIONS=suppressions=asan/lsan.supp ASAN_OPTIONS=heap_profile=0:unmap_shadow_on_exit=1:detect_invalid_pointer_pairs=1:symbolize=1:detect_leaks=1:dump_instruction_bytes=1:print_suppressions=0 src/tarantool -e 'local s = "local a; a="..string.rep("{", 400) loadstring(s) os.exit()'
Segmentation fault (core dumped)
```

In scope of #8870 the new CMake configuration option `FIBER_STACK_SIZE` has been introduced. I changed the stack size to 1Mb and everything works fine.

```
$ CC=clang-11 CXX=clang++-11 cmake . -DCMAKE_BUILD_TYPE=Debug -DENABLE_WERROR=ON -DENABLE_ASAN=ON -DENABLE_UB_SANITIZER=ON -DENABLE_FUZZER=ON -DFIBER_STACK_SIZE=1Mb
<...>
$ cmake --build . --parallel 8
<...>
$ ASAN=ON LSAN_OPTIONS=suppressions=asan/lsan.supp ASAN_OPTIONS=heap_profile=0:unmap_shadow_on_exit=1:detect_invalid_pointer_pairs=1:symbolize=1:detect_leaks=1:dump_instruction_bytes=1:print_suppressions=0 src/tarantool -e 'local s = "local a; a="..string.rep("{", 400) loadstring(s) os.exit()'
$
```

I propose to close the ticket if it works for you. Otherwise, please clarify your concerns regarding the original report.

**Comment by Buristan** [comment]
Closed via #8870.
